### PR TITLE
feat(desktop): make Project Chats chat-first and searchable

### PR DESCRIPTION
**Pull request by xyhStruggler**
## Summary

- keep the Figma Projects sessions overview as the canonical project landing, then open a selected session in the contextual workbench
- make the detailed Project Chats surface chat-first with title search, running/waiting/done/failed filters, and bounded cursor-based Load more
- preserve explicit Board selection, new-chat selection during asynchronous refreshes, shared inspector behavior, persistent project context, and the existing Gateway/IPC workspace contract
- ignore known first-page replays for exhausted collections while surfacing legitimately new items below the cap, keep full bounded projections terminal until explicit refresh, and cancel queued page loads when the runtime/account generation changes
- extract the pure project thread rail model before growing the composition component
- ordinary-merge the latest `main` through `74a1b223ebae9f78d69803fc62c21bed05e6f31e`, including #1282, #1253, #1293, #1288, #1252, and #1296

This adapts the strongest T3 Code navigation behavior to Matrix's existing architecture. It does not import T3 runtime packages, persistence, provider adapters, or Electron backend code; the PR diff remains limited to the Project chat/search/pagination implementation and its tests.

Linear: [MAT-345](https://linear.app/matrix-os/issue/MAT-345/pr-1-make-project-chats-navigation-chat-first-and-searchable)  
Parent: [MAT-344](https://linear.app/matrix-os/issue/MAT-344/implement-a-t3-code-inspired-project-chat-experience)

## Verification

Exact reviewed head: `024bc664547b98b716809cb3d5c985d0c9831e65`

- expanded Projects/Chat/navigation suite: 446/446 passed across 30 affected test files
- focused project workspace store suite: 22/22 passed
- built Electron project sequence (auth → task → Git → project session/chat/Board/review): 4/4 passed
- Gateway/kernel/platform/proxy/edge-router/observability and Desktop typechecks: passed
- `bun run check:patterns`: 0 violations; 5 repository-wide warnings
- `bun run build:desktop`: passed
- `git diff --check origin/main...HEAD`: passed

The local dependency tree had been regenerated before this integration. Its missing root test links initially caused import failures and a duplicate `react-resizable-panels` resolution that bypassed a test mock; aligning those generated links restored the unchanged suite to green without any tracked source or test accommodation.

### Source of truth

- Matrix Gateway's project workspace projection remains authoritative for tasks and project/task thread summaries.
- Desktop stores only bounded projections and the existing runtime-scoped view preference; no second thread registry is introduced.

### Lock/transaction scope

- No database write, lock, or transaction behavior changes.
- Pagination is read-only and reuses the existing validated Gateway route and cursor contract.

### Acceptable orphan states

- No durable objects are created by search, filtering, view selection, or pagination.
- A failed Load more keeps the last bounded workspace projection visible through the existing stale-while-revalidate error path.

### Auth source of truth

- Existing authenticated Gateway and trusted Desktop IPC boundaries are unchanged.
- The renderer receives only the existing validated project workspace projection.

### Deferred scope

- thread archive/rename, pin, snooze, and lifecycle persistence
- server-side full-text search and global all-project navigation
- inspector execution-root provenance work planned for PR 2 and PR 3
- any T3 server, SQLite store, provider adapter, or Electron backend reuse

### CODE REVIEW

**Comment by xyhStruggler** [comment]
@greptileai Please review the exact current head 063241e6fbad16290f703a4eaadb6d44ed5601f0. Focus on correctness, regressions, state ownership, accessibility, and test coverage. Do not infer approval from the Draft state.

File: desktop/src/renderer/src/stores/project-workspaces.ts [hunk 1]
```diff
@@ -111,7 +121,70 @@ function isStaleLoad(projectId: string, runtimeGeneration: number, generation: n
...
+    updatedAt: next.updatedAt,
+  };
+}
+
+async function performWorkspaceLoad(projectId: string, append = false): Promise<void> {
+  const previousWorkspace = useProjectWorkspaces.getState().entries[projectId]?.workspace ?? null;
+  const shouldAppend = append && previousWorkspace !== null;
+  if (shouldAppend && !canLoadMoreWorkspace(previousWorkspace)) return;
   const runtimeGeneration = captureRuntimeGeneration();
   const generation = nextGeneration(projectId);
```
[the remark this answers is not in the document]
Comment (xyhStruggler) [reply desktop/src/renderer/src/stores/project-workspaces.ts]: Fixed in 1988b402f. Project workspace mutations now serialize per project within the same runtime scope, so overlapping pagination requests merge each successful page in order. Runtime/account changes invalidate the old queue without blocking the new scope. Covered by the public store regression and the existing runtime-generation race tests; the focused suite is 154/154 green.

File: desktop/src/renderer/src/stores/project-workspaces.ts [hunk 2]
```diff
@@ -111,7 +121,70 @@ function isStaleLoad(projectId: string, runtimeGeneration: number, generation: n
...
+  const projectThreadCursor = pageCursor(workspace.projectThreads);
+  const taskThreadCursor = pageCursor(workspace.taskThreads);
+  return {
+    projectId,
+    ...(taskCursor ? { taskCursor } : {}),
+    taskLimit: pageLimit(workspace.tasks),
+    ...(projectThreadCursor ? { projectThreadCursor } : {}),
+    projectThreadLimit: pageLimit(workspace.projectThreads),
+    ...(taskThreadCursor ? { taskThreadCursor } : {}),
+    taskThreadLimit: pageLimit(workspace.taskThreads),
```
[the remark this answers is not in the document]
Comment (xyhStruggler) [reply desktop/src/renderer/src/stores/project-workspaces.ts:164]: Fixed in 1988b402f. Pagination requests now include a cursor only for collections whose page still reports hasMore; exhausted collections keep their bounded items for deduplication but no longer send stale cursors. The canonical request-shape regression is green.

File: desktop/src/renderer/src/features/project/ProjectTab.tsx [hunk 3]
```diff
@@ -58,7 +58,7 @@ export function ProjectViewSwitch({
  * panels (git DAG, files, terminal) to this header/segmented structure.
  */
 export default function ProjectTab({ projectSlug, active }: { projectSlug: string; active: boolean }) {
-  const view = useProjectView((s) => s.entries[projectSlug]?.view ?? "board");
+  const view = useProjectView((s) => s.entries[projectSlug]?.view ?? DEFAULT_PROJECT_VIEW);
```
[the remark this answers is not in the document]
Comment (xyhStruggler) [reply desktop/src/renderer/src/features/project/ProjectTab.tsx:65]: Fixed in 1988b402f. Project view hydration now reconciles view and selectedThreadId independently using bounded field mutation revisions, preserving a persisted Board choice when chat auto-selection lands during the async read. The public store hydration regression is green.

File: desktop/src/renderer/src/features/project/ProjectThreadList.tsx [hunk 4]
```diff
@@ -341,13 +286,30 @@ export function ProjectThreadList({
 
         {isEmpty && status !== "loading" ? (
           <p className="px-2 py-2 text-[11px]" style={{ color: "var(--text-tertiary)" }}>
-            No chats yet. Start one with the + button above.
+            {query.trim() || statusFilter !== "all"
+              ? "No chats match these filters."
+              : "No chats yet. Start one with the + button above."}
```
[the remark this answers is not in the document]
Comment (xyhStruggler) [reply desktop/src/renderer/src/features/project/ProjectThreadList.tsx:297]: Fixed in 1988b402f. The filtered empty state now scopes its claim to loaded chats while pagination remains and explicitly directs the operator to Load more. The rendered component regression verifies both the copy and action.

**Comment by xyhStruggler** [comment]
@greptileai Please re-review the exact current head 1988b402fdd4f0b1cf704eccc2e920ee56c2ecf3 after the four fixes. Confirm correctness, concurrency behavior, hydration semantics, accessibility, and regression coverage. Report the confidence score and any remaining blocking findings.

**Comment by xyhStruggler** [comment]
@Nima-Naderi This PR is ready for your review. Greptile is 5/5 on exact head 1988b402fdd4f0b1cf704eccc2e920ee56c2ecf3, all review threads are resolved, and the current checks have no pending or failing jobs.

**Comment by xyhStruggler** [comment]
[codex] Integrated the latest `main` through #1288 with an ordinary merge at exact head `b9b1df522dd311cc220f017335173970a260c0b3`. The affected Projects/Desktop/Gateway suites (334/334), post-integration Projects + Terminal/navigation suite (304/304), built Electron project sequence (4/4), additional MAT-348/provider checks (2/2), Terminal E2E (4/4), typecheck, Desktop build, pattern scan (0 violations), and diff check pass.

A prior broad stateful operator run was not fully green (11/18), so this handoff intentionally claims only the exact affected sequences rerun after the latest merge.

@greptileai Please re-review this exact current head for integration regressions, state ownership, concurrency, accessibility, and coverage, and report the confidence score plus any blocking findings.

**Comment by xyhStruggler** [comment]
[codex] Re-integrated the latest `main` through #1252 and #1296 with an ordinary merge. Exact head is now `de7fe75052f1c18e18b089b12d29f93c28554b96`; the expanded Projects/Chat/navigation suite is 442/442 green, the built Electron project sequence is 4/4 green, all affected package/Desktop typechecks pass, the Desktop production build passes, the pattern scan reports 0 violations, and the diff check passes.

The earlier local failures were generated dependency-link resolution only; no tracked source or test changes were made to accommodate them.

@greptileai Please discard the superseded-head result and review this exact current head for integration regressions, state ownership, concurrency, accessibility, and coverage. Report the confidence score and any remaining blocking findings.

**Review by Nima-Naderi — APPROVED** [review]

File: desktop/src/renderer/src/stores/project-workspaces.ts [hunk 5]
```diff
@@ -115,10 +132,74 @@ function isStaleLoad(projectId: string, runtimeGeneration: number, generation: n
...
```
[the remark this answers is not in the document]
Comment (xyhStruggler) [reply desktop/src/renderer/src/stores/project-workspaces.ts:164]: [codex] Fixed in 49ac42f7c. Exhausted collections now keep their terminal hasMore/cursor state when the shared workspace route restarts that collection while paginating a sibling. The new public-store regression reproduces the resurrection and is green.

File: desktop/src/renderer/src/stores/project-workspaces.ts [hunk 6]
```diff
@@ -182,13 +267,23 @@ async function performWorkspaceLoad(
...
+  const token = loadQueueTokens[projectId] ?? Symbol(projectId);
+  loadQueueTokens[projectId] = token;
+  const previous = activeLoadPromises[projectId];
+  // Start an idle queue immediately. Besides avoiding an unnecessary tick,
+  // this preserves the public contract that invoking refresh has dispatched
+  // its request before the caller can switch runtime/account scope.
+  const work = previous
+    ? previous.then(async () => {
+        if (loadQueueTokens[projectId] !== token) return;
+        await performWorkspaceLoad(projectId, options);
```
[the remark this answers is not in the document]
Comment (xyhStruggler) [reply desktop/src/renderer/src/stores/project-workspaces.ts:307]: [codex] Fixed in 49ac42f7c. Queued loads capture the runtime generation at enqueue time and are cancelled before starting after an account/runtime change. The new regression verifies that the queued request never reaches the Gateway.

**Comment by xyhStruggler** [comment]
@greptileai please review the exact current head 49ac42f7c4e4a853fe40e833e3608ff92631339c.

File: desktop/src/renderer/src/stores/project-workspaces.ts [hunk 7]
```diff
@@ -115,10 +132,79 @@ function isStaleLoad(projectId: string, runtimeGeneration: number, generation: n
...
+
+function mergePage<T extends { id: string }>(
+  current: WorkspacePage<T>,
+  next: WorkspacePage<T>,
+): WorkspacePage<T> {
+  // The shared Gateway route always returns every collection. For a
+  // collection that is already exhausted, a pagination request for another
+  // collection therefore contains a restarted first page. Keep the terminal
+  // state authoritative instead of resurrecting its cursor/hasMore flag.
+  if (!current.hasMore) return current;
```
[the remark this answers is not in the document]
Comment (xyhStruggler) [reply desktop/src/renderer/src/stores/project-workspaces.ts]: [codex] Fixed in 69cb72269957c112ad6c9fab26ce77bdf6045a26. Exhausted collections now ignore known first-page replays but prepend legitimately new IDs and adopt the returned cursor so pagination can continue. TDD: the new regression was red before the change; the focused store suite is 21/21, the expanded Projects/Chat/navigation suite is 445/445, and the built Electron sequence is 4/4.

**Comment by xyhStruggler** [comment]
@greptileai please review the exact current head 69cb72269957c112ad6c9fab26ce77bdf6045a26.

File: desktop/src/renderer/src/stores/project-workspaces.ts [hunk 8]
```diff
@@ -115,10 +132,89 @@ function isStaleLoad(projectId: string, runtimeGeneration: number, generation: n
...
+  // collection therefore contains a restarted first page. Ignore a pure
+  // replay, but surface genuinely new first-page items and reopen pagination
+  // so additional new items can be discovered without a full refresh.
+  if (!current.hasMore) {
+    const prepended = next.items.filter((item) => !knownIds.has(item.id));
+    if (prepended.length === 0) return current;
+    return {
+      items: [...prepended, ...current.items].slice(0, MAX_WORKSPACE_PAGE_ITEMS),
+      hasMore: next.hasMore,
+      ...(next.nextCursor ? { nextCursor: next.nextCursor } : {}),
```
[the remark this answers is not in the document]
Comment (xyhStruggler) [reply desktop/src/renderer/src/stores/project-workspaces.ts]: [codex] Fixed in 9b96c02024351e9c4af25a27706c31117533d6e7. When a newly discovered first-page item would overflow a full exhausted projection, Desktop now restarts that collection from the authoritative first page and cursor; the next load has capacity and pages the displaced prior items back in. The new 100-item regression was red before the change. Focused store tests are 22/22, expanded Projects/Chat/navigation tests are 446/446, and the built Electron sequence is 4/4.

**Comment by xyhStruggler** [comment]
@greptileai please review the exact current head 9b96c02024351e9c4af25a27706c31117533d6e7.

File: desktop/src/renderer/src/stores/project-workspaces.ts [hunk 9]
```diff
@@ -115,10 +132,101 @@ function isStaleLoad(projectId: string, runtimeGeneration: number, generation: n
...
+    const mergedItems = [...prepended, ...current.items];
+    // A full bounded projection cannot prepend without making its displaced
+    // tail unreachable. Restart from the authoritative first page instead;
+    // its cursor leaves capacity to page the prior items back in.
+    if (mergedItems.length > MAX_WORKSPACE_PAGE_ITEMS) {
+      return {
+        items: next.items.slice(0, MAX_WORKSPACE_PAGE_ITEMS),
+        hasMore: next.hasMore,
+        ...(next.nextCursor ? { nextCursor: next.nextCursor } : {}),
+        limit: next.limit,
```
[the remark this answers is not in the document]
Comment (xyhStruggler) [reply desktop/src/renderer/src/stores/project-workspaces.ts]: [codex] Fixed in 024bc664547b98b716809cb3d5c985d0c9831e65. A full exhausted projection now remains terminal and stable when a sibling response reveals an overflow item; explicit refresh is the only operation that replaces that authoritative bounded page. Below the 100-item cap, legitimately new IDs are still prepended. The cap-boundary regression was red before this change. Focused store tests are 22/22, expanded Projects/Chat/navigation tests are 446/446, and the built Electron sequence is 4/4.

**Comment by xyhStruggler** [comment]
@greptileai please review the exact current head 024bc664547b98b716809cb3d5c985d0c9831e65.

**Comment by xyhStruggler** [comment]
[codex] @Nima-Naderi We are accelerating this merge to support tonight’s canary build and validation window, so we are not waiting for an additional human review. Exact head `024bc664547b98b716809cb3d5c985d0c9831e65` includes current main, has Greptile 5/5 with zero unresolved review threads, and all current GitHub checks pass or are intentionally skipped; 22/22 focused store tests, 446/446 expanded Projects/Chat/navigation tests, built Electron project sequence 4/4, Desktop typecheck and build, pattern scan (0 violations), and diff check are green. Yuhan has explicitly authorized landing qualifying Desktop/T3 PRs now and will also explain the timing directly.
